### PR TITLE
Using Logistic Regression in PyTorch to Identify Handwritten Digits

### LINKED ISSUE #4591: [Languages] Using Logistic Regression in PyTorch to Identify Handwritten Digits
#### Proposed title of article
Using Logistic Regression in PyTorch to identify handwritten digits
#### Proposed article introduction
When there are several explanatory variables, logistic regression is used to calculate the odds ratio. With the exception that the response variable is binomial, the approach is quite similar to multiple linear regression. The impact of each variable on the odds ratio of the observed event of interest is the end outcome.
The statistical method of logistic regression allows us to predict a binary outcome given a set of independent factors. Now we'll look at how to do this in PyTorch, as well as how to use Logistic Regression in PyTorch to identify handwritten digits from the MNIST dataset.
#### Key takeaways 
- Knowing The Dataset
- Build The Neural Network
-  Adjusting Weights
- Core Training Process
#### References
N/A
### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by masese911**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?
This fixes #4591

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
Please always use the template provided by us.
I edited your initial comment and added the template. Check what matches your article.

**Comment by masese911** [comment]
@lalith1403 Could you please look at this PR whenever you are free. Thanks.

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

**Comment by masese911** [comment]
@lalith1403 @hectorkambow It's been 12days already. What is the status of this PR?

**Comment by lalith1403** [comment]
Will review it by eod.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+topics: [Languages]
+excerpt_separator: <!--more-->
+images:
+
+ - url: /engineering-education/implementing-gan-from-scratch/hero.jpg
+   alt: Implementing GANs example image
+---
+
+### Introduction
+The odds ratio can be calculated using logistic regression when there are multiple explanatory variables. The approach is quite similar to multiple linear regression, except that the response variable is a binomial. An event's odds ratio is the final result of each variable's influence on the odds ratio.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:19]: This is an abrupt introduction to the article. Also, when you introduce a concept, explain it. Odds ratio may not be familiar to all.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+### Table of contents
+- [Library functions and objects](#library-functions-and-objects)
+- [Defining our hyperparameters](#defining-our-hyperparameters)
+- [Building The Neural Network](#building-the-neural-network)
+- [References ](#references )
+- [Conclusion](#conclusion)
+ 
+### Library functions and objects
+Importing Library functions and objects into your code is necessary once your installation is complete.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:35]: Library need not be capitalized. Avoid capitalizing until it is a proper noun.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+Here are the functions and objects to import:
+
+- The **torch.nn** module: Contains code for the required model.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:38]: What model are you talking about? Doesn't it contain abstract implementations of a components of a nn?

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+- The **torchvision.datasets**: Includes MNIST dataset of handwritten digits that we shall be using here.
+- The **torchvision.transforms**: We shall be using it here to tt from images to PyTorch tensors.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:40]: what is tt?

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+Let us define our hyperparameters:
+```Python
+input_size = 784
+num_classes = 10
+num_epochs = 5
+batch_size = 100
+learning_rate = 0.001
+
+```
+The image size that we will use for our image will be 28*28. This means that our input size will be 784. In addition, there are ten digits in this. Thus we can generate ten alternative results. In this way, number_classes will be set to ten. In addition, we'll run the full dataset through five iterations of training. Finally, we will train in small batches of 100 images each so that the software does not crash because of memory overflow.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:76]: Any variable or keyword that you refer to from the code, use \` \` to wrap the keyword.
For example: \` number_classes \` will look like `number_classes`

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+After that, we'll lay out our model in the following manner. In this section, we'll define the forward pass after setting up our model as a subclass of **torch.tn.Module**. As we are building the code, we will not need to mention the softmax in the **forward()** function because it will be determined internally during each forward run.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:78]: For bold, use double underscores and for italics, single underscores. Just a common practice

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	def forward(self, x):
+		out = self.linear(x)
+		return out
+```
+Now that our class has been established, we may create an instance of it.
+```Python
+model = LogisticRegression(input_size, num_classes)
+```
+### Building The Neural Network
+Our loss function and optimizer are now set. As specified in the hyperparameter above, we'll use the cross-entropy loss and the stochastic gradient descent algorithm with a reading rate of 0.001 for the optimizer.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:94]: Possibly add links to CE loss and SGD

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      self.fch2 = tn.Linear(50, 10)
+   def forward(self, b):
+      b = tF.relu(tF.max_pool2d(self.c1(b), 2))
+      b = tF.relu(tF.max_pool2d(self.c2_drop(self.c2(b)), 2))
+      b = b.view(-1, 320)
+      b = tF.relu(self.fch1(b))
+      b = tF.dropout(b, trai=self.training)
+      b = self.fch2(b)
+      return tF.log_softmax(b)
+```
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:113]: Explain the code.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+To get the network and the optimizer up and running;
+```python
+criterion = tn.CrossEntropyLoss()
+network = NeuralNetwork()
+opti = torch.optim.SGD(newmodel.parameters(), lr = reading_rate)
+```
+We're ready to begin training now. Resetting all gradients to 0 will be the first step here, followed by a forward pass, the loss calculation, backpropagation, and the updating of all weights.
+```python
+# Training the Model
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:122]: Explain the dataloader part of it.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+	outputs = model(images)
+	_, predicted = torch.max(outputs.data, 1)
+	total += labels.size(0)
+	correct += (predicted == labels).sum()
+
+print('Accuracy of the model on the 10000 test images: % d %%' % (
+			100 * correct / total))
+
+```
+To put it another way, if you followed the instructions exactly, you would have an accuracy rate of 91%, which is significantly lower than the current best model, which makes use of a different sort of neural network architecture.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:152]: This is an incorrect way to put it. Accuracy is a measure of the data and data-preprocessing model capacity. So instead of giving them a number they should attain, mention that by following the steps, you got 91%.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+You can run the whole code [here](https://colab.research.google.com/drive/1eL6a4_QxAZxqLV83vJOsLkPF09hYwThn?usp=sharing)
+
+### References 
+1. [To see whole code click here](https://colab.research.google.com/drive/1eL6a4_QxAZxqLV83vJOsLkPF09hYwThn?usp=sharing)
+2. [PyTorch](https://drive.google.com/drive/folders/0B41Zbb4c8HVyUndGdGdJSXd5d3M?resourcekey=0-s90CYmIbmbqbO1Mvtwmlog)
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:158]: What link is this? I see a drive link.

**Review by lalith1403 — CHANGES_REQUESTED** [review]
There are several changes required. Please go through each one of them diligently and let me know. Explanations are required in several places to improve readability.

**Comment by masese911** [comment]
@lalith1403 thanks
On them right away

**Comment by masese911** [comment]
@lalith1403 Done.

**Comment by hectorkambow** [comment]
@masese911 

Lets please be sure that we are submitting a final & polished article that is ready to published. To help prevent any possible back and forth as much as possible. (as there may have been other issues that were not pointed out).
🚀

This help improve any review time.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+- [Conclusion](#conclusion)
+ 
+### Library functions and objects
+Importing library functions and objects into your code is necessary once your installation is complete.
+Here are the functions and objects to import:
+
+- The **torch.nn** module: Contains code thatwe will be using in our model.
+- The **torchvision.datasets**: Includes MNIST dataset of handwritten digits that we shall be using here.
+- The **torchvision.transforms**: We shall be using it to transform from images to PyTorch tensors.
+- The **torch.autograd**: This will be used to define our tensors.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:40]: please check the functionality of autograd. It is not used to define the tensors.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+input_size = 784
+num_classes = 10
+num_epochs = 5
+batch_size = 100
+learning_rate = 0.001
+
+```
+The image size that we will use for our image will be 28*28. This means that our input size will be 784. In addition, there are ten digits in this. Thus we can generate ten alternative results. In this way, `number_classes` will be set to ten. In addition, we'll run the full dataset through five iterations of training. Finally, we will train in small batches of 100 images each so that the software does not crash because of memory overflow.
+
+After that, we'll lay out our model in the following manner. In this section, we'll define the forward pass after setting up our model as a subclass of __torch.tn.Module__. As we are building the code, we will not need to mention the softmax in the __forward()__ function because it will be determined internally during each forward run.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:77]: torch.nn.Module

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+	def forward(self, x):
+		out = self.linear(x)
+		return out
+```
+Now that our class has been established, we may create an instance of it.
+```Python
+model = LogisticRegression(input_size, num_classes)
+```
+### Building The Neural Network
+Our loss function and optimizer are now set. As specified in the hyperparameter above, we'll use the [cross-entropy loss](https://en.wikipedia.org/wiki/Cross_entropy) and the [stochastic gradient descent algorithm](https://en.wikipedia.org/wiki/Stochastic_gradient_descent) with a reading rate of 0.001 for the optimizer.
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:93]: hyperparameters *

Comment (lalith1403) [reply content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:93]: What is reading rate?

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md (hunk 13 above)
Comment (lalith1403) [reply content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:77]: Isn't it usually imported as nn? Correct me if I am wrong here. If you want to proceed with tn, explicitly mention what tn is.

File: content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+      b = self.fch2(b)
+      return tF.log_softmax(b)
+```
+To get the network and the optimizer up and running;
+```python
+criterion = tn.CrossEntropyLoss()
+network = NeuralNetwork()
+opti = torch.optim.SGD(newmodel.parameters(), lr = reading_rate)
+```
+We're ready to begin training now. Resetting all gradients to 0 will be the first step here, followed by a forward pass, the loss calculation, backpropagation, and updating all weights. Dataloader will load the individual batches. We set the gradients to zero using `optimizer.zero_grad().` The `backward()` call will now collect a new set of gradients which we propagate back into each of the network’s parameters using the optimizer.step().
```
Comment (lalith1403) [inline content/articles/using-logistic-regression-in-pytorch-to-identify-handwritten-digits/index.md:121]: `optimizer.step()`

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Several errors still exist in the article. When we suggest errors, we do not review all the errors—only the immediately noticeable ones. I found additional errors in the second round of review (grammatical, technical explanations). 

We want our articles to be technically accurate. Please re-check the article. Also, the article is significantly light on content. On the other hand, the classification of the MNIST dataset using PyTorch is extensively covered. You may introduce several concepts, such as building several models and using them on MNIST. The choice is up to you. 

Make the changes and let me know. Thank you. I will be closing the PR for now.

File: content/authors/masese-gavin/index.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
+---
+title: Masese Gavin
+type: authors
+github: https://github.com/masese911
+---
+ I’m currently learning computer science. I’m looking to collaborate on software development. I play games whenever i am free.
```
Comment (lalith1403) [inline content/authors/masese-gavin/index.md:6]: Look at other author descriptions. This is an excellent opportunity for people to recognize you. Make sure you add all the required details.

**Review by lalith1403 — COMMENTED** [review]
Add details that you feel fit. It is an excellent opportunity to make new connections, let the world know about you and build a portfolio. Thank you.

**Comment by masese911** [comment]
@lalith1403 
Hello.
I have the changes with me, just waiting to be committed. Please reopen so that I can push them and ensure I have revised the article once more so that there will be no more revisions. Thank you in advance.

**Comment by masese911** [comment]
@lalith1403 should I reopen another PR?

**Comment by masese911** [comment]
> Several errors still exist in the article. When we suggest errors, we do not review all the errors—only the immediately noticeable ones. I found additional errors in the second round of review (grammatical, technical explanations).
> 
> We want our articles to be technically accurate. Please re-check the article. Also, the article is significantly light on content. On the other hand, the classification of the MNIST dataset using PyTorch is extensively covered. You may introduce several concepts, such as building several models and using them on MNIST. The choice is up to you.
> 
> Make the changes and let me know. Thank you. I will be closing the PR for now.

I have the changes. Reopen so that I can commit them. Thanks.
